HDDS-6391. ClientVersions and DatanodeVersions class to define an enum with version and description

fapifta opened the pull request:
## What changes were proposed in this pull request?

As discussed in the proposal for HDDS-6390, this PR is to change int constants defined in ClientVersions and DatanodeVersions class to enums, providing the same functionality with an int version number, and a description field.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-6391

## How was this patch tested?
Same tests via CI should cover the changes. No new tests were added.

adoroszlai commented on hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/ClientVersion.java:
```diff
@@ -20,19 +20,34 @@
 
-  // this should always point to the latest version
-  public static final int CURRENT_VERSION = VERSION_HANDLES_UNKNOWN_DN_PORTS;
+  private int version;
+  private String description;
```
```suggestion
  private final int version;
  private final String description;
```

fapifta commented on hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/DatanodeVersion.java:
```diff
@@ -20,16 +20,30 @@
+  SEPARATE_RATIS_PORTS_AVAILABLE(1, "Version with separated Ratis port.");
 
-  // this should always point to the latest version
-  public static final int CURRENT_VERSION = SEPARATE_RATIS_PORTS_AVAILABLE;
```
I would argue that Enum.values() creates a new array each time. values is an implicit method, and according to the Enum<E extends Enum> class documentation, it relies on Class<?>#getEnumConstants() method to return the enum values array.
Internally this uses a volatile array, and the implicit values() method returns just a reference to that array which is initialized lazily on first access inside the Class object.

So all in all, in the latest() method we store that reference internally, and then use it twice. I haven't searched for written evidence, but I would bet JIT optimize this under the hood and does not create a new reference in the stack just inlines the access and uses the reference in the class object.

adoroszlai replied:
```
  @Test
  public void enumValues() {
    ClientVersion[] v1 = ClientVersion.values();
    ClientVersion[] v2 = ClientVersion.values();
    assertSame(v1, v2);
  }
```

```
AssertionError: expected same:<[Lorg.apache.hadoop.ozone.ClientVersion;@2ea227af> was not:<[Lorg.apache.hadoop.ozone.ClientVersion;@4386f16>
```

fapifta replied:
Hmm... so sad I did not went to try this out, but apparently you are right... I will change this shortly. Thank you for noting this!

adoroszlai reviewed (commented):
Thanks @fapifta for updating the patch.  I don't have any further comments.

errose28 reviewed (commented):
Thanks for the changes @fapifta. As purely a rename that maintains the original functionality this looks fine to me, however I think the implementation is confusing on a few fronts. This is not anything new your patch introduced, but I think these changes provide us an opportunity to clear things up.

I am mostly concerned about confusion between `DatanodeVersion` and `HDDSLayoutFeature#layoutVersion`. Since we are just passing around ints right now, I think the following mistake is easy to make:
```java
public void foo(int datanodeVersion); // Expects something from the DatanodeVersion enum.
...
foo(HDDSLayoutFeature.DATANODE_SCHEMA_V2.layoutVersion()); // Wrong, but an honest mistake that is hard to catch.
```

I see a few ways we can prevent this:

1. I think `DatanodeVersion` might become confusing to other devs, since it is a pretty generic name and we already have "layout versions" in the upgrade framework. Maybe we could add a qualifier to the name, something like `{Shared,External,Advertised,Broadcasted}DatanodeVersion` to indicate that it is not internal, like layout features, but meant to be shared with other components to make decisions when communicating. Better javadoc for the enum would help too.

2. Use a new common interface for `ClientVersion`, `DatanodeVersion`, `OzoneManagerVersion`, etc. This is what the upgrade framework does: `HDDSLayoutFeature` and `OMLayoutFeature` both implement `LayoutFeature`. This provides a few things:
    - A way for the code to show that these versions are used for related tasks.
    - Shows that the versions are distinct from the set of upgrade related layout versions.
    - Provides a central hub linking all these related version classes.
        - Going to the interface and finding implementors in an IDE shows that each component has its own set of versions used in a similar way.

3. Use the enum type instead of an int as soon as the version is taken from the proto. It would be the client/server side translators' responsibility to translate from the int to the enum. This enforces use of the correct version scheme at compile time. I admit some of [this discussion](https://github.com/apache/ozone/pull/3155/files#r818408119) went over my head, but I think we could use `DatanodeVersion#CURRENT` to refer to the latest one, so I'm not sure why `DatanodeVersion#CURRENT_VERSION` would be needed.

Having to pass `ClientVersion.CURRENT_VERSION` when constructing protos on the server side is a bit confusing, but it looks like an unfortunate consequence of using the same protos for internal processing and client communication. I think passing the enum type instead of an int will help reduce errors, but that may be the best we can do.

adoroszlai commented:
Good points @errose28.

* I also wanted to suggest using the enum type as parameter instead of `int`, but we have to consider the "newer than any versions this component is aware of (i.e. has enum constants for)" case.  Do we create a `FUTURE_VERSIONS(Integer.MAX_VALUE)` enum constant for that case?  At least we should have an overloaded version.

* I have already found a case where the wrong version is being passed unintentionally (HDDS-6418).

* Regarding `{Shared,External,Advertised,Broadcasted}DatanodeVersion`: I'm not sure why plain `DatanodeVersion` would imply that it is internal.

* `ComponentVersion` might be a good name for the common interface.

* The discussion we had about `CURRENT_VERSION` is mostly implementation-specific.  The point is that we should have a constant rather than a (very slightly expensive) method.

> Having to pass `ClientVersion.CURRENT_VERSION` when constructing protos on the server side is a bit confusing, but it looks like an unfortunate consequence of using the same protos for internal processing and client communication.

It might also indicate a case where client version is ignored (and could result in compatibility problem).

errose28 commented:
Thanks for the reply @adoroszlai 

> but we have to consider the "newer than any versions this component is aware of (i.e. has enum constants for)" case. Do we create a FUTURE_VERSIONS(Integer.MAX_VALUE) enum constant for that case?

Yeah this is tricky. At least by providing a `FUTURE_VERSION` enum value, we are acknowledging the existence of future versions. There's nothing with the int that suggests to people using it they may see a value larger than `CURRENT_VERSION`. I think adding this enum constant would be a good idea, along with maybe a helper in the enum to go from the int version to the enum value. This would return the `FUTURE_VERSION` enum member when an int larger than `CURRENT_VERSION` is given.

> - Regarding {Shared,External,Advertised,Broadcasted}DatanodeVersion: I'm not sure why plain DatanodeVersion would imply that it is internal.
> - ComponentVersion might be a good name for the common interface.

I was mostly just looking to add some sort of descriptor to distinguish this version from layout versions. As you suggested, "component version" may be a good name. I think something like `DatanodeComponentVersion implements ComponentVersion` would serve this purpose. This gives us unique terminology to use in explanations, for example "Layout version is used to track the format of on disk data. Component version is used to determine compatibility between components"

fapifta commented:
Thank you for the reviews, I went down deeper into the code that is touched by this PR, and I found a couple of issues that we have to address, and I think I have kind of a plan.

So, currently we have the problem listed by Attila in HDDS-6418.
Besides that I have found that the container operation client uses the client version, but in that case the Ozone Manager is the client, so that has to be changed if we have committed the OzoneManagerVersion class.
There is the Balancer, there is the ReplicationManager which are also using the ClientVersion now, and I am 99% sure that they should not.
There is PipelineManager which also provides the ClientVersion.CURRENT_VERSION to a toProto method, which is certainly not right similarly as it is not right in HDDS-6418.

On the other hand, you are right, the problem when a new client is connecting is real, and a FUTURE_VERSION or something has to be added. I think the version number of the FUTURE_VERSION constant should be -1, and that should be the 0th ordinal in the enum, so that does not leads extra code during latest version calculation.
The version difference check is a bit tricky this way though, but it can be done for example this way:
(clientversion & Integer.MAX_VALUE) > serverversion -> newer client

I am adding utility methods to compare the versions into the base class that is proposed for the enums, and refresh the PR shortly.

Because of the problems with the usage of ClientVersion and DatanodeVersion, I would like to suggest to go and commit this PR first, and in the meantime I will file follow up JIRAs to review the usage of these, and then while we are unblocked to continue with FSO and EC and start to use these changes, while we also can aim to switch to enum usage wherever it is appropriate in the current code where we need to fix things anyway. What do you think?

adoroszlai commented:
@fapifta compile fails because recently merged HDDS-6379 added more usage of `ClientVersions.CURRENT_VERSION` in `TestPipelinePlacementPolicy`.

adoroszlai commented:
License header is also missing from `ComponentVersion`.

fapifta commented:
I hope I have fixed all the issues, will be back to check in an hour, thanks for checking @adoroszlai

errose28 reviewed (commented):
I'm ok with -1 being used to represent future versions, but I'm not sure about the current implementation surrounding it:

1. Putting the compare method as default in the interface pushes incorrect version comparison validation (ex. datanode <-> client) to runtime instead of compile time. Since there will be a finite number of version enums that may be compared many times in the future, Some code duplication in their compare methods is a better tradeoff IMO.
2. By putting it in the implementor, not the interface, the version does not need to be passed in. This removes confusion over which side is which in the params.
3. There are now 3 different ways to compare versions:
    1. {Datanode,Client}Version#compareTo
        - Implemented by default for enums
     2. int comparison with {Datanode,Client}Version#version.
     3. ComponentVersion#compare
         - With the -1 value and FUTURE_VERSION being the first enum entry, this is the only correct comparison. It is also the least intuitive as the other two are standard among java types and this one is not.

Let me know what you think of this proposal to fix the above issues:
1. Make FUTURE_VERSION the last enum value.
    - This can be enforced with a unit test to prevent errors when more versions are added.
2. Get rid of the `ComponentVersion#version` method.
    - {to,from}ProtoValue methods can be added to translate the enum type to/from the wire. I don't think any other places would need the literal value given by the version.

With this approach the enum's default comparator implementation is correct, and the only possible way to compare versions.

Sorry if the critiques sounds pedantic, but I think the errors around existing version usage mentioned above in this PR and in #3173 indicate that a more relaxed approach will lead to confusion, misuse, and bugs later on. I think if we construct this feature right, we can prevent future devs from making version mistakes almost completely at compile time.

fapifta commented:
Hi @errose28,

thank you for the review and for the suggestions, I think the suggested design is way better than we had initially, so I have implemented the suggested changes, and fixed the comparisons I could find.

Also thank you @adoroszlai for the discussions before, and the reviews.

I think the reasonable next step is to switch to use the enum wherever we can, and just translate it to int over the wire in a separate JIRA, though before doing so, I would like to take a quick experiment with an idea to reduce the number of times we need to pass on the version because of DatanodeDetails#toProto. Based on the experiment results, I will create the follow up JIRAs tomorrow.
After that and when this PR goes in you can expect an update on the OzoneManagerVersion related PR, and then I will also post the PR that introduces the usage of the OzoneManagerVersion.

adoroszlai commented on hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/DatanodeVersion.java:
```diff
@@ -0,0 +1,65 @@
+  }
+
+  public static DatanodeVersion fromProtoValue(int value) {
+    DatanodeVersion[] versions = DatanodeVersion.values();
+    if (value >= versions.length || value < 0) {
+      return FUTURE_VERSION;
+    }
+    return versions[value];
```
Minor suggestion for improvement, while we wait for @errose28 to take a look.

```java
  private static final Map<Integer, DatanodeVersion> BY_PROTO_VALUE =
      Arrays.stream(DatanodeVersion.values())
        .collect(toMap(DatanodeVersion::toProtoValue, identity()));

  public static DatanodeVersion fromProtoValue(int value) {
    return BY_PROTO_VALUE.getOrDefault(value, FUTURE_VERSION);
  }
```

This allows gaps in `version` number, and avoids the `values()` call for each lookup.

errose28 reviewed (approved):
Thanks for the updates @fapifta LGTM. Attila's suggestion looks like a good addition too.
